docs/source/clickhouse/schema_design.rst

```diff
@@ -0,0 +1,35 @@
+ClickHouse Schema Design Best Practices
+=======================================
+
+Columns based on dictionary data
+--------------------------------
+
+ClickHouse is a columnar datastore, and at run-time it loads columns on-demand
+based on the columns referenced in the query (both the columns ``SELECT`` ed
+and those part of the ``WHERE`` clause). The ability to store different columns independently
+and not load them for every row for every query is part of the performance advantage that
+ClickHouse provides over a traditional RDBMBS (like PostgreSQL).
+
+Commonly, a data schema contains a flexible key:value pair mapping (canonically: ``tags``) and stores that
+data in a column that contains two ``Nested`` arrays where the first array contains the keys
+of the dictionary and the second array contains the values. This works well when
+your dataset design gives you the ability to filter based on other attributes to a small
+number of rows and the flexibility of completely arbitrary keys and values are a real requirement.
+Often, however, a ClickHouse user is interested in rows that contain a specific tag key or a
+specific ``key=value`` pair and in this case one is often best-served by creating a top-level
+column and duplicating (we often call it "promoting" in code) the data.
```

Comment:
Storing duplicates has storage cost associated with it. Ideally, when a specific array element is being promoted to its own column, you would stop storing it in the array. But the problem is that the promotion does not help with data which is already persisted on Clickhouse. So the promotion would start happening from the day promotion code is merged. Then to use only the new columns in a consistent manner you would have to wait for 90 days data to be written. Then you can stop writing the duplicate data.

Reply:
I'll change the wording around duplication, that makes sense